framework/sandstone.cpp

```diff
@@ -1042,6 +1047,87 @@ int num_packages() {
 }
 
 namespace {
+class BarrierDeviceSchedule : public DeviceSchedule
+{
+public:
+    void device_reschedule() override {
+        auto on_completion = [&]() noexcept {
+            std::unique_lock lock(groups_mutex);
+
+            int group_idx = thread_num / members_per_group;
+            GroupInfo *group = groups[group_idx];
+
+            // Rotate cpus vector so reschedule group members to a different cpu
+            std::rotate(group->next_cpu.begin(), group->next_cpu.begin() + 1, group->next_cpu.end());
+            lock.unlock();
+
+            // Reschedule group members
+            for (int i=0; i<members_per_group; i++) {
+                if(!pin_thread_to_logical_processor(LogicalProcessor(cpu_info[group->next_cpu[i]].cpu_number), group->pid[i]))
+                    log_debug("Failed to reschedule %d (%lld) to CPU %d", thread_num, pthread_self(), cpu_info[group->next_cpu[i]].cpu_number);
+            }
+        };
+
+        std::unique_lock lock(groups_mutex);
+
+        // Initialize groups on first run
+        if (groups.size() == 0) {
+            groups.reserve(num_cpus()/members_per_group);
+            for (int i=0; i<num_cpus(); i+=members_per_group) {
+                groups.push_back(new GroupInfo(members_per_group, on_completion));
+            }
+        }
+
+        // Fill thread info if not done already
+        int group_idx = thread_num / members_per_group;
+        GroupInfo *group = groups[group_idx];
+        int thread_info_idx = thread_num % members_per_group;
+        if (group->pid[thread_info_idx] == 0) {
+            group->pid[thread_info_idx] = getpid();
+            group->next_cpu[thread_info_idx] = thread_num;
+        }
+
+        lock.unlock();
+
+        // Wait on proper barrier
+        group->barrier->arrive_and_wait();
+        return;
+    }
+
+    void finish_reschedule() override {
+        // When thread finishes, unsubscribe it from barrier
+        // this avoid partners deadlocks
+        int group_idx = thread_num / members_per_group;
+        GroupInfo *group = groups[group_idx];
+
+        // Remove thread info from groups
+        std::unique_lock lock(groups_mutex);
+        int thread_info_idx = thread_num % members_per_group;
+        // group->next_cpu.erase(group->next_cpu.begin() + thread_info_idx); // Should I remove the cpus as well?
```

Comment:
Good question. We can leave the CPU there, so the remaining threads would run there. But we may not want to do that if the execution has failed for this thread... that's probably a bad device.

You can query if we have failed using
```c++
    sApp->test_thread_data(thread_number)->has_failed()
```

Explain the logic in a comment in the source, for future reference.